Async API

superquadratic opened the pull request:
Make the whole API async. E.g. instead of `db.run(sql)`, we now use `db.runAsync(sql)` which returns a `WinJS.Promise`. The old synchronous API has been removed. Support for working with statements directly has also been dropped for now.

Closes #7

superquadratic committed:
Make opening a database connection async

superquadratic committed:
Move opening brackets to the same line in C++

- More consistent with JavaScript code
- Works better with lambdas

superquadratic committed:
Make statement preparation async

As a result, most of the JavaScript Database functions have become async, too.

superquadratic committed:
Disabled most of the spec for the refactoring

superquadratic committed:
Move Database::runAsync() and oneAsync() to C++

superquadratic committed:
Move allAsync() to C++

superquadratic committed:
Fix capturing of parameter arrays

Capturing IVector and alike in lambdas doesn't seem to be supported yet. Fixed
by copying the parameters into an std::vector.

superquadratic committed:
Support binding of NULL parameters in C++

superquadratic committed:
Implement eachAsync() in C++

superquadratic committed:
Implement mapAsync() in terms of C++ eachAsync()

superquadratic committed:
Reenable all specs

superquadratic committed:
Removed JS Statement class

The statement functionality is now implemented in C++ in SQLite3::Statement.
For now, statements cannot be used directly in JavaScript, only implicitly
using Database::runAsync() etc.

superquadratic committed:
Remove custom typeof implementation

We can do without it now that we no longer handle edge cases like typeof null
on the JavaScript side.

superquadratic committed:
Refactor SQLite error handling

superquadratic committed:
Refactor error handling for SQLite steps

Moved error handling from the call sites into the Step() wrapper. Simplifies
the call sites (Run(), One()) and adds error handling to other places (All(),
Each()).

superquadratic committed:
Fix behavior of oneAsync() when no row is returned

It now returns null instead of a row full of null values.

superquadratic committed:
Group and order specs by tested function

superquadratic committed:
Spec result of allAsync() for empty queries

superquadratic committed:
Ignore insertion order in specs

superquadratic committed:
Allow chaining of runAsync() and eachAsync()

pke commented on SQLite3JS/js/SQLite3.js:
```diff
@@ -226,44 +110,54 @@
   );
 
-  GroupDataSource = WinJS.Class.derive(WinJS.UI.VirtualizedDataSource,
+   GroupDataSource = WinJS.Class.derive(WinJS.UI.VirtualizedDataSource,
```
Is this indention intentionally?

milgner commented:
+1 - this is good stuff! :)

superquadratic committed:
Update README

- Updated SQLite3JS example to use async API
- Stripped down SQLite3 component description and removed the example

The JavaScript API is what I think most people will use this project for, so
it should be featured prominently in the README. We do not have any experience
with using the component in C#, so we just mention it shortly. Actually, there
might be better options for C#/.NET, e.g. sqlite-net.
